July and August recap (#654)
Roblox/luau: docs/_posts/2022-08-29-luau-recap-august-2022.md

---
layout: single
title:  "Luau Recap: August 2022"
---

Luau is our new language that you can read more about at [https://luau-lang.org](https://luau-lang.org).

[Cross-posted to the [Roblox Developer Forum](https://devforum.roblox.com/t/luau-recap-august-2022/).]

## Tables now support `__len` metamethod

See the RFC [Support `__len` metamethod for tables and `rawlen` function](https://github.com/Roblox/luau/blob/master/rfcs/len-metamethod-rawlen.md) for more details.

> With generalized iteration released in May, custom containers are easier than ever to use. The only thing missing were the fact tables didn't respect `__len`.

Reviewer: ```suggestion
With generalized iteration released in May, custom containers are easier than ever to use. The only thing missing was the fact that tables didn't respect `__len`.
```

The function `my_cool_lower` is a function `<a...>(t: t1) -> a... where t1 = {+ lower: (t1) -> a... +}`. 

```lua
function my_cool_lower(t)
    return t:lower()
end
```

Even though `t1` is a table type, we know `string` is a subtype of `t1` because `string` also has `lower` which is a subtype of `t1`'s `lower`, so this call site now type checks.

```lua
local s: string = my_cool_lower("HI")
```

## Other analysis improvements

* `string.gmatch`/`string.match`/`string.find` may now return more precise type depending on the patterns used
* Fix a bug where type arena ownership invariant could be violated, causing stability issues
* Fix a bug where internal type error could be presented to the user
* Fix a false positive with optionals & nested tables
* Fix a false positive in non-strict mode when using generalized iteration
* Improve autocomplete behavior in certain cases for `:` calls
* Fix minor inconsistencies in synthesized names for types with metatables
* Fix autocomplete not suggesting globals defined after the cursor
* Fix DeprecatedGlobal warning text in cases when the global is deprecated without a suggested alternative

> * Fix an off-by-one error in type error text for incorrect use of string.format

Reviewer 2: ```suggestion
* Fix an off-by-one error in type error text for incorrect use of `string.format`
```

## Other runtime improvements

* Comparisons with constants are now significantly faster when using clang as a compiler (10-50% gains on internal benchmarks)
* When calling non-existent methods on tables or strings, `foo:bar` now produces a more precise error message
* Improve performance for iteration of tables
* Slightly reduce interpreter I$ footprint
* Fix a bug with negative zero in vector components when using vectors as table keys

> * Compiler can now constant fold builtins under -O2

Reviewer 2: ```suggestion
* Compiler can now constant fold builtins under -O2, for example `string.byte("A")` is compiled to a constant
```

* Compiler can model the cost of builtins for the purpose of inlining/unrolling
* Local reassignment i.e. `local x = y :: T` is free iff neither `x` nor `y` is mutated/captured

> * Improve debug.traceback performance by 1.15-1.75x depending on the platform

Reviewer 2: ```suggestion
* Improve `debug.traceback` performance by 1.15-1.75x depending on the platform
```

> * Fix a corner case with table assignment semantics when key didn't exist in the table and __newindex was defined: we now use Lua 5.2 semantics and call __newindex, which results in less wasted space, support for NaN keys in __newindex path and correct support for frozen tables

Reviewer 2: ```suggestion
* Fix a corner case with table assignment semantics when key didn't exist in the table and `__newindex` was defined: we now use Lua 5.2 semantics and call `__newindex`, which results in less wasted space, support for NaN keys in `__newindex` path and correct support for frozen tables
```

* Reduce parser C stack consumption which fixes some stack overflow crashes on deeply nested sources

> * Improve performance of bit32.extract/replace when width is implied (~3% faster chess)

Reviewer 2: ```suggestion
* Improve performance of `bit32.extract`/`replace` when width is implied (~3% faster chess)
```

> * Improve performance of bit32.extract when field/width are constants (~10% faster base64)

Reviewer 2: ```suggestion
* Improve performance of `bit32.extract` when field/width are constants (~10% faster base64)
```

> 

Reviewer 2: ```suggestion
* `string.format` now supports a new format specifier, `%*`, that accepts any value type and formats it using `tostring` rules
```
